### PR TITLE
Safari tests and proxy support

### PR DESCRIPTION
**Pull request by adomasven**
Adds:
- Testing support. Open test runner via preferences when built with Debug flags
- Proxy support without automatic proxy detection

Notes:
- No testing via travis or selenium. Currently not possible
- When loaded via the Extension Builder unicode characters are not recognised (hopefully a packaged extension works fine, but I have no way to test) image
- No Reload via Proxy menu yet. Needs additional work on the popover, will add later.

Addresses https://github.com/zotero/zotero-connectors/issues/135

### CODE REVIEW

**Comment by dstillman** [comment]
> When loaded via the Extension Builder unicode characters are not recognised (hopefully a packaged extension works fine, but I have no way to test)

I can check in a package, but did you try adding a BOM to the file?

File: src/browserExt/test/testSetup.js [hunk 1]
```diff
@@ -115,6 +118,9 @@ Zotero.initDeferred.promise.then(function() {
 				}
 			}
 		});
+		Zotero.Background.getTabById = async function(tabId) {
```
[the remark this answers is not in the document]
Comment (adomasven) [reply src/browserExt/test/testSetup.js]: All of chrome and safari APIs use `tabId`, `extensionUrl`, etc, so this is consistent.

[the remark this answers is not in the document]
Comment (dstillman) [reply src/browserExt/test/testSetup.js]: They use `tabId` for variable names in documentation. We always use capital letters in functions. And your `getExtensionUrl()` below is literally calling `browser.extension.getURL()`.

File: src/common/api.js [hunk 2]
```diff
@@ -246,8 +246,8 @@ Zotero.API = new function() {
 					"Zotero-API-Version": "3"
 				}
 			};
-			return Zotero.HTTP.request("POST", url, options).then(xhr => xhr.responseText).catch(function(e) {
-				if (askForAuth && xmlhttp.status === 403) {
+			return Zotero.HTTP.request("POST", url, options).then(xhr => xhr.responseText, function(e) {
```
[the remark this answers is not in the document]
Comment (adomasven) [reply src/common/api.js]: Yeah, but may as well change it. This is more correct in terms of what we're expecting to catch.

[the remark this answers is not in the document]
Comment (dstillman) [reply src/common/api.js]: It's just less consistent and harder to scan for no reason. It'll still throw some hypothetical error accessing `xhr.responseText` regardless.

**Comment by adomasven** [comment]
> > When loaded via the Extension Builder unicode characters are not recognised (hopefully a packaged extension works fine, but I have no way to test)
> 
> I can check in a package, but did you try adding a BOM to the file?

UTF-8 BOM is set on the built files

**Comment by dstillman** [comment]
>>  I can check in a package, but did you try adding a BOM to the file?
>
> UTF-8 BOM is set on the built files

Not sure what you're referring to. There's no BOM in `build/safari.safariextension/proxy.js` for me.

**Comment by adomasven** [comment]
> > UTF-8 BOM is set on the built files
> 
> Not sure what you're referring to. There's no BOM in `build/safari.safariextension/proxy.js` for me.

```bash
adomas@paprika-macbook:~/zotero/connectors
→ file -I ./build/safari.safariextension/proxy.js 
./build/safari.safariextension/proxy.js: text/x-c; charset=utf-8
```

**Comment by adomasven** [comment]
> It looks like this has the unchecked "Automatically associate new hosts" problem — not sure if that would come in a merge or if this undoes it.

See https://github.com/zotero/zotero-connectors/commit/61bf118c261e7091c0d0a2ebe493b8e6197f9fee

**Comment by dstillman** [comment]
> adomas@paprika-macbook:~/zotero/connectors
> → file -I ./build/safari.safariextension/proxy.js 
> ./build/safari.safariextension/proxy.js: text/x-c; charset=utf-8

That's not a file with a BOM. UTF-8 != BOM

**Comment by dstillman** [comment]
I'm not actually seeing this problem jstor.org, though, even without a BOM. On what site did you see this? I wondering if it's specific to the encoding of the page.

**Comment by adomasven** [comment]
> That's not a file with a BOM. UTF-8 != BOM

Hmm, so it seems. Either way according to [stackoverflow](https://stackoverflow.com/a/2223926/3199106)

> The UTF-8 BOM is a sequence of bytes (EF BB BF) that allows the reader to identify a file as being encoded in UTF-8.
> 
> Normally, the BOM is used to signal the endianness of an encoding, but since endianness is irrelevant to UTF-8, the BOM is unnecessary.
> 
> According to the Unicode standard, the BOM for UTF-8 files is not recommended

I'll look into this tomorrow, it might have to do with the website encoding

**Comment by adomasven** [comment]
Happens on JSTOR for me

**Comment by adomasven** [comment]
Ok, fixed with https://github.com/zotero/zotero-connectors/pull/183/commits/587e7de8130f8d79b7ca354cdec8ecf58c5eed34

**Comment by dstillman** [comment]
This has some conflicts now.

**Comment by adomasven** [comment]
Good to merge otherwise?

**Comment by dstillman** [comment]
I want to test it again once the conflicts are resolved.

**Comment by dstillman** [comment]
I'm getting a (rather vague) init error when it tries to reset translators automatically:

> [JavaScript Error: "init@safari-extension://org.zotero.zoteroconnectorforsafari-8LAYR367YV/b8d7a78f/translators.js:87:29
update@safari-extension://org.zotero.zoteroconnectorforsafari-8LAYR367YV/b8d7a78f/translators.js:253:45
_handleResponse@safari-extension://org.zotero.zoteroconnectorforsafari-8LAYR367YV/b8d7a78f/repo.js:178:29
safari-extension://org.zotero.zoteroconnectorforsafari-8LAYR367YV/b8d7a78f/repo.js:141:19
promiseReactionJob@[native code]"]

It points to [this line](https://github.com/zotero/zotero-connectors/blob/e9e9752dc1fa6969cdfc1fe8cd922b505a9b46a9/src/common/translators.js#L87), and (I think) proxies aren't coming through from the client when it occurs.

It went away after a few reloads, and then proxies came through, but I reproduced it with

```
localStorage['pref-translatorMetadata'] = JSON.stringify(JSON.parse(localStorage['pref-translatorMetadata']).slice(0, 10))
```
and toggling the extension off and on.

**Comment by adomasven** [comment]
Wouldn't affect [proxy initialisation](https://github.com/adomasven/zotero-connectors/blob/fd9322b966350af73bd436380e3cb947b0cbdcb2/src/common/zotero.js#L127-L134) in any way, translator/repo init occurs [asynchronously](https://github.com/adomasven/zotero-connectors/blob/4e884b51bec980a621053ec95e545ebb056fd06b/src/common/repo.js#L38). 

And you can certainly reproduce the logged error by only keeping 10 translators - that's [exactly](https://github.com/adomasven/zotero-connectors/commit/1af9f5d94619c9cb5612e955eea4869eb5e7b603) what it's meant to safeguard against. In safari the logged error is rather cryptic, but it's good enough to figure out what's happening if it crops up in error reports, so I'm not worried.

**Comment by dstillman** [comment]
OK, something else may have been breaking proxy initialization, then, but unless I can reproduce it, we can ignore for now.

> And you can certainly reproduce the logged error by only keeping 10 translators - that's exactly what it's meant to safeguard against. In safari the logged error is rather cryptic, but it's good enough to figure out what's happening if it crops up in error reports, so I'm not worried.

Well I know what it's doing — that's how I reproduced it. But you're saying that the logged message is just logError() messages not coming through properly? If so, we should create a ticket for that, if we think it's fixable.

This can be rebased and merged.

**Comment by adomasven** [comment]
Merged in https://github.com/zotero/zotero-connectors/commit/249bba2c40ace472c4c9b619b3f1437f4adbd71c

**Pull request closed without merging**